ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,499 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
                            QUOTA_LIMITS_HARDWS_MAX_DISABLE)
 
 /* PRIVATE FUNCTIONS *******************************************************/
 
-/*
- * Private helper to charge the specified process quota.
- * Returns STATUS_QUOTA_EXCEEDED on quota limit check failure.
- * Updates QuotaPeak as needed for specified quota type in PS_QUOTA_TYPE enum.
- * Notes: Conceptually translation unit local/private.
+/**
+ * @brief
+ * Returns pool quotas back to the Memory Manager
+ * when the pool quota block is no longer being
+ * used by anybody.
+ *
+ * @param[in] QuotaBlock
+ * The pool quota block which quota resources
+ * are to be sent back.
+ *
+ * @return
+ * Nothing.
+ */
+VOID
+NTAPI
+PspReturnQuotasOnDestroy(
+    _In_ PEPROCESS_QUOTA_BLOCK QuotaBlock)
+{
+    ULONG PsQuotaTypeIndex;
+    SIZE_T QuotaToReturn;
+
+    /* Make sure that the quota block is not plain garbage */
+    ASSERT(QuotaBlock);
+
+    /* Loop over the Process quota types */
+    for (PsQuotaTypeIndex = PsNonPagedPool; PsQuotaTypeIndex < PsQuotaTypes; PsQuotaTypeIndex++)
+    {
+        /* Terminate the loop if we reach here */
+        if (PsQuotaTypeIndex == PsPageFile)
+        {
+            break;
+        }
+
+        /* The amount needed to return to Mm is the limit and return fields */
+        QuotaToReturn = QuotaBlock->QuotaEntry[PsQuotaTypeIndex].Limit + QuotaBlock->QuotaEntry[PsQuotaTypeIndex].Return;
+        MmReturnPoolQuota(PsQuotaTypeIndex, QuotaToReturn);
+    }
+}
+
+/**
+ * @brief
+ * Releases some of excess quotas in order to
+ * attempt to free up some resources. This is
+ * done primarily in case the Memory Manager
+ * fails to raise the quota limit.
+ *
+ * @param[in] QuotaType
+ * Process pool quota type.
+ *
+ * @param[out] ReturnedQuotas
+ * A pointer to the returned amount of quotas
+ * back to Memory Manager.
+ *
+ * @return
+ * Nothing.
+ *
+ * @remarks
+ * The function releases excess paged or non
+ * paged pool quotas. Page file quota type is
+ * not permitted.
+ */
+VOID
+NTAPI
+PspReturnExcessQuotas(
+    _In_ PS_QUOTA_TYPE QuotaType,
+    _Outptr_ PSIZE_T ReturnedQuotas)
+{
+    PLIST_ENTRY PspQuotaList;
+    PEPROCESS_QUOTA_BLOCK QuotaBlockFromList;
+    SIZE_T AmountToReturn = 0;
+
+    /*
+     * Loop over the quota block lists and reap
+     * whatever quotas we haven't returned which
+     * is needed to free up resources.
+     */
+    for (PspQuotaList = PspQuotaBlockList.Flink;
+         PspQuotaList != &PspQuotaBlockList;
+         PspQuotaList = PspQuotaList->Flink)
+    {
+        /* Gather that list */
+        QuotaBlockFromList = CONTAINING_RECORD(PspQuotaList, EPROCESS_QUOTA_BLOCK, QuotaList);
+
+        /*
+         * Gather any unreturned quotas and cache
+         * them to a variable. If there's nothing
+         * to return back then there's nothing to do...
+         */
+        AmountToReturn += InterlockedExchangeAddSizeT(&QuotaBlockFromList->QuotaEntry[QuotaType].Return, 0);
```

Comment:
That would force me to write a macro like `InterlockedExchangeSizeT` or similar since there are `InterlockedExchange` and `InterlockedExchange64` written separately in **interlocked.h** header. Plus this trick of doing `InterlockedAdd(x, 0)` was originally suggested by you before.